src/System.Drawing.Common/src/System/Drawing/Icon.Unix.cs

```diff
@@ -709,77 +709,163 @@ private void InitFromStreamWithSize(Stream stream, int width, int height)
+                    using (var bihReader = new BinaryReader(new MemoryStream(buffer)))
+                    {
+                        uint headerSize = bihReader.ReadUInt32();
 
-            //iconDir = new IconDir ();
-            iconDir.idReserved = reader.ReadUInt16();
-            if (iconDir.idReserved != 0) //must be 0
-                throw new System.ArgumentException("Invalid Argument", nameof(stream));
+                        // Vista 256x256 icons points directly to a PNG bitmap
+                        // 256x256 icons are decoded as 0x0 (width and height are encoded as BYTE)
+                        // and we ignore them just like MS does (at least up to fx 2.0)
+                        // Added: storing data so it can be saved back
+                        if ((ide.width == 0 && ide.height == 0) || headerSize == (uint)(('P' << 8) | ('N' << 16) | ('G' << 24) | 0x89))
+                        {
+                            IconDump id = new IconDump
+                            {
+                                data = buffer
+                            };
+                            imageData[i] = id;
+                            iconDir.idEntries[i].ignore = true;
+                        }
+                        else
+                        {
+                            valid = true;
+                            BitmapInfoHeader bih = new BitmapInfoHeader
+                            {
+                                biSize = headerSize,
+                                biWidth = bihReader.ReadInt32(),
+                                biHeight = bihReader.ReadInt32(),
+                                biPlanes = bihReader.ReadUInt16(),
+                                biBitCount = bihReader.ReadUInt16(),
+                                biCompression = bihReader.ReadUInt32(),
+                                biSizeImage = bihReader.ReadUInt32(),
+                                biXPelsPerMeter = bihReader.ReadInt32(),
+                                biYPelsPerMeter = bihReader.ReadInt32(),
+                                biClrUsed = bihReader.ReadUInt32(),
+                                biClrImportant = bihReader.ReadUInt32()
+                            };
+
+                            IconImage iidata = new IconImage
+                            {
+                                iconHeader = bih
+                            };
 
-            iconDir.idType = reader.ReadUInt16();
-            if (iconDir.idType != 1) //must be 1
-                throw new System.ArgumentException("Invalid Argument", nameof(stream));
+                            // Read the number of colors used and corresponding memory occupied by
+                            // color table. Fill this memory chunk into rgbquad[]
+                            // number of colors = 2 ^ biBitCount
+                            int numColors = 0;
+                            if (bih.biBitCount < 24)
```

Comment:
Previously it only accepted 2, 16, or 256 colors, and now you accept up to 16M ? 
It looks like Windows does, but it also does more checks like against biClrUsed, which isn't happening here.
I would leave this code the same unless and until we need to change it.

Reply:
Oh yes you’re right somehow when reading the docs I missed that 16 bitCount was still supported without having to use biClrUsed. Would you bring back the switch just for readability? Or just change this if to <= 8

Reply:
Oh yes you’re right, great catch! somehow when reading the docs I thought that 16 bitCount was still supported without having to use biClrUsed. Would you bring back the switch just for readability? Or just change this if to <= 8

Reply:
Oh yes you’re right, great catch! somehow when reading the docs I thought that 16 bitCount was still supported without having to use biClrUsed. Would you bring back the switch just for readability? Or just change this if to <= 8

Reply:
Probably the switch since otherwise it would accept new values like 128 that windows does not

Reply:
Ok. Makes sense then.